### PR TITLE
Exemplar implementation

### PR DESCRIPTION
**Pull request by ramonza**
Fixes: #873

### CODE REVIEW

**Comment by ramonza** [comment]
@odeke-em still waiting on the spec to be merged (https://github.com/census-instrumentation/opencensus-specs/pull/179) but what do you think of the approach?

File: stats/doc.go [hunk 1]
```diff
@@ -21,35 +21,49 @@ aggregate the collected data, and export the aggregated data.
...
 Measurement is a data point to be collected for a measure. For example,
 for a latency (ms) measure, 100 is a measurement that represents a 100ms
-latency event. Users collect data points on the existing measures with
+latency event. Measurements are created from measures with
 the current context. Tags from the current context are recorded with the
 measurements if they are any.
 
-Recorded measurements are dropped immediately if user is not aggregating
-them via views. Users don't necessarily need to conditionally enable/disable
+Recorded measurements are dropped immediately if no views are registered for then.
```
Comment (odeke-em) [inline stats/doc.go]: s/then/them/

but we could improve that sentence by saying
```go
Recorded measurements are dropped immediately if they don't have any associated views.
```

File: stats/view/aggregation_data.go [hunk 2]
```diff
@@ -89,18 +92,35 @@ func (a *SumData) equal(other AggregationData) bool {
...
-	Mean            float64   // mean of the distribution
-	SumOfSquaredDev float64   // sum of the squared deviation from the mean
-	CountPerBucket  []int64   // number of occurrences per bucket
-	bounds          []float64 // histogram distribution of the values
+	Count           int64   // number of data points aggregated
+	Min             float64 // minimum value in the distribution
+	Max             float64 // max value in the distribution
+	Mean            float64 // mean of the distribution
+	SumOfSquaredDev float64 // sum of the squared deviation from the mean
+	// Deprecated: Replaced by Count in Buckets.
```
Comment (odeke-em) [inline stats/view/aggregation_data.go]: Perhaps let's put this field at the bottom and a blank space before it to draw attention to it like this:
```go
     bounds []float64 // histogram distribution of the values
     Buckets []Bucket
     
     // Deprecated: Replaced by Count in Buckets.
     CountPerBucket []int64 // Number of occurances per bucket
```

File: stats/view/aggregation_data.go [hunk 3]
```diff
@@ -139,26 +159,61 @@ func (a *DistributionData) addSample(f float64) {
...
 
-	for i, b := range a.bounds {
-		if f < b {
-			a.CountPerBucket[i]++
-			return
+	// Heuristic to pick the "better" exemplar: first keep the one with a
+	// sampled trace attachment, if neither have a trace attachment, pick the
+	// one with more attachments.
+	_, haveTraceID := e.Attachments["TraceID"]
+	if haveTraceID || len(e.Attachments) >= len(b.Exemplar.Attachments) {
```
Comment (odeke-em) [inline stats/view/aggregation_data.go]: I believe we meant to use `>` instead of `>=` otherwise it'll always create a new Exempler if `!haveTraceID && len(e.Attachments) == len(b.Exemplar.Attachments)`

File: stats/view/aggregation_data.go [hunk 4]
```diff
@@ -139,26 +159,61 @@ func (a *DistributionData) addSample(f float64) {
...
 }
 
 func (a *DistributionData) clone() AggregationData {
+	c := *a
 	counts := make([]int64, len(a.CountPerBucket))
 	copy(counts, a.CountPerBucket)
-	c := *a
 	c.CountPerBucket = counts
+	buckets := make([]Bucket, len(a.Buckets))
+	copy(buckets, a.Buckets)
```
Comment (odeke-em) [inline stats/view/aggregation_data.go]: This only changes the slices but the underlying ExemplarData per bucket is the same
since the ExemplarData pointer is unchanged so this is a shallow copy. Copying CountsPerBucket copies int64s which are just values.

File: trace/exemplar.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+	ExemplarKeyTraceOptions = "TraceOptions"
+)
+
+func init() {
+	exemplar.RegisterAttachmentExtractor(attachSpanContext)
+}
+
+func attachSpanContext(ctx context.Context, attachments map[string]string) map[string]string {
+	span := FromContext(ctx)
+	sc := span.SpanContext()
```
Comment (odeke-em) [inline trace/exemplar.go]: This might result in a nil dereference on the first span(if SpanContext doesn't check that it is nil), but anyways we don't use it if span == nil. Please move it below after we've checked `if span == nil`

File: trace/exemplar_test.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+		t.Fatalf("bucket.Count = %d; want %d", got, want)
+	}
+	exemplar := bucket.Exemplar
+	if exemplar == nil {
+		t.Fatal("Expected exemplar")
+	}
+	if got, want := exemplar.Value, 1.5; got != want {
+		t.Fatalf("exemplar.Value = %v; got %v", got, want)
+	}
+	if _, ok := exemplar.Attachments["TraceID"]; !ok {
```
Comment (odeke-em) [inline trace/exemplar_test.go]: Could we please use ExemplarTraceID and ExemplerSpanID, etc.. as the keys in these tests?

File: exemplar/extractor.go [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+	"time"
+)
+
+// Exemplar represents an abstract set of supplemental information associated with
+// a measurement, for the purpose of providing an example of the type of
+// situation that resulted in the measurement.
+// For example, a trace exemplar contains key-value pairs representing a
+// SpanContext and points to an example span that resulted in a given
+// measurement.
+type Exemplar struct {
```
Comment (odeke-em) [inline exemplar/extractor.go]: Am curious why we have two very similar structs `Exemplar` and `ExemplarData` which vary by only one field `Value`, but also this definition of Exemplar seems only suited to traces yet we expect them to work for both traces and metrics.

File: stats/view/aggregation_data.go [hunk 8]
```diff
@@ -89,18 +92,35 @@ func (a *SumData) equal(other AggregationData) bool {
...
+	SumOfSquaredDev float64 // sum of the squared deviation from the mean
+	// Deprecated: Replaced by Count in Buckets.
+	CountPerBucket []int64   // number of occurrences per bucket
+	bounds         []float64 // histogram distribution of the values
+	Buckets        []Bucket
+}
+
+// ExemplarData is an example data point associated with each bucket.
+type ExemplarData struct {
+	Value       float64           // the value that was recorded
```
Comment (odeke-em) [inline stats/view/aggregation_data.go]: Nice! I was actually thinking to comment about making this a float64, but you already did it.

File: stats/view/aggregation_data.go [hunk 9]
```diff
@@ -139,26 +159,61 @@ func (a *DistributionData) addSample(f float64) {
...
-		a.CountPerBucket[0]++
+func (a *DistributionData) addToBucket(val float64, ex *exemplar.Exemplar) {
+	var (
+		count  *int64
+		bucket *Bucket
+	)
+
+	for i, b := range a.bounds {
+		if val < b {
+			count = &a.CountPerBucket[i]
```
Comment (odeke-em) [inline stats/view/aggregation_data.go]: I now wonder why bounds, Bucket aren't of type `map[int]float64`.
I believe it would save on:
* Potential runtime crashes from out-of-bounds index access with slices
* Having to understand that the zeroth bucket is for the case when we don't have any bounds
* Save on the pointer manipulation
* Save on upfront memory allocations e.g. if we had a program using lots of views with DistributionData
* [index] access would automatically create zero values

To illustrate this point example this function would simply become
```go
func (a *DistributionData) addToBucket(val float64, ex *exemplar.Exemplar) {
        valBucketIndex := len(a.bounds)
        for i, b := range a.bounds {
                if val < b {
                        valBucketIndex = i 
                        break
                }
        }

        bucket := a.Buckets[valBucketIndex]
        bucket.Count += 1
        a.CountPerBucket[valBucketIndex] = bucket.Count
        bucket.maybeRetainExemplar(val, ex)
}
```

Although perhaps it is too late to change the types of Count* and Bucket* but it is something to keep in mind for future designs :)

File: stats/doc.go (hunk 1 above)
Comment (ramonza) [reply stats/doc.go]: I think that the current phrasing is better since it is very specific in saying "registered" (which is the name of the function) while "associated" is more vague.

File: stats/view/aggregation_data.go (hunk 3 above)
Comment (ramonza) [reply stats/view/aggregation_data.go]: Yeah, the idea is that we replace the existing exemplar with the newer one if they are equally "good" under the assumption that newer is better.

File: stats/view/aggregation_data.go (hunk 4 above)
Comment (ramonza) [reply stats/view/aggregation_data.go]: Yeah but we don't expect anyone to mutate the map at this stage (it's only mutated when the exemplary is first created) so I think shallow copying in this case is fine.

File: trace/exemplar_test.go (hunk 6 above)
Comment (ramonza) [reply trace/exemplar_test.go]: These are the keys from the spec. I think since they map is already inside the exemplar, repeating the work "exemplar" doesn't really add anything.

File: exemplar/extractor.go (hunk 7 above)
Comment (ramonza) [reply exemplar/extractor.go]: Done

File: stats/view/aggregation_data.go (hunk 4 above)
Comment (ramonza) [reply stats/view/aggregation_data.go]: I created a type for attachments and added this as a note on the type godoc.

File: exemplar/extractor.go [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package exemplar implements support for exemplars. Exemplars are additional
+// data associated with each measurement.
+//
+// Their purpose it to provide an example of the kind of thing
+// (request, RPC, trace span, etc.) that resulted in that measurement.
+package exemplar
```
Comment (rakyll) [inline exemplar/extractor.go]: Is this a public API or should it be internal for now?

What are the public use cases?

Comment (ramonza) [reply exemplar/extractor.go]: The Exemplar type is needed for exporters and the AttachmentExtractor is meant to be an extensible mechanism (see comments on the spec PR about ways to extract attachments from Tracestate).

File: exemplar/extractor.go [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+	extractors = append(extractors, e)
+}
+
+// NewFromContext extracts exemplars from the given context.
+// Each registered AttachmentExtractor (see RegisterAttachmentExtractor) is called in an
+// unspecified order to add attachments to the exemplar.
+func AttachmentsFromContext(ctx context.Context) Attachments {
+	var a Attachments
+	for _, extractor := range extractors {
+		a = extractor(ctx, a)
```
[the remark this answers is not in the document]
Comment (ramonza) [reply exemplar/extractor.go]: We may have more than one extractor return valid attachments. For example, if there are tags in the context as well as a sampled trace.

File: exemplar/extractor.go [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+	"time"
+)
+
+const (
+	KeyTraceID   = "trace_id"
+	KeySpanID    = "span_id"
+	KeyPrefixTag = "tag:"
+)
+
+// Exemplar is an example data point associated with each bucket.
```
[the remark this answers is not in the document]
Comment (ramonza) [reply exemplar/extractor.go]: Yes, it is only available for distribution aggregation. I will make that more explicit.

File: exemplar/extractor.go [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+type Exemplar struct {
+	Value       float64     // the value that was recorded
```
[the remark this answers is not in the document]
Comment (ramonza) [reply exemplar/extractor.go]: This is how it is in the metrics proto: https://github.com/census-instrumentation/opencensus-proto/blob/a74b002d33b4471aff7288afdb41393f6ff02ca1/src/opencensus/proto/metrics/v1/metrics.proto

File: stats/view/aggregation_data.go [hunk 14]
```diff
@@ -17,14 +17,16 @@ package view
...
+	"go.opencensus.io/exemplar"
 )
 
 // AggregationData represents an aggregated value from a collection.
 // They are reported on the view data during exporting.
 // Mosts users won't directly access aggregration data.
 type AggregationData interface {
 	isAggregationData() bool
-	addSample(v float64)
+	addSample(e *exemplar.Exemplar)
```
[the remark this answers is not in the document]
Comment (ramonza) [reply stats/view/aggregation_data.go]: I agree this is a bit odd. The reason for this diff (removing the v argument) is that it is repeated inside the exemplar.

Regarding the presence of exemplars in the core APIs (it being a top-level package), I am open to moving it under "stats" or even "stats/view". But then we would have a direct dependency from the trace package to a package under "stats".

File: stats/view/aggregation_data.go [hunk 15]
```diff
@@ -89,18 +91,30 @@ func (a *SumData) equal(other AggregationData) bool {
...
+	bounds []float64 // histogram distribution of the values
+
+	// Deprecated: Replaced by Count in Buckets.
+	CountPerBucket []int64 // number of occurrences per bucket
+}
+
+// Bucket represents a single bucket value in a distribution.
+type Bucket struct {
+	Count    int64              // Number of recorded measurements falling into this bucket.
+	Exemplar *exemplar.Exemplar // Exemplar of this bucket.
```
[the remark this answers is not in the document]
Comment (ramonza) [reply stats/view/aggregation_data.go]: Yes, see Bucket in: https://github.com/census-instrumentation/opencensus-proto/blob/master/src/opencensus/proto/metrics/v1/metrics.proto

File: stats/view/aggregation_data.go (hunk 2 above)
Comment (ramonza) [reply stats/view/aggregation_data.go]: ok, I think this makes sense especially given that most Exporters won't care about exemplars.

**Comment by ramonza** [comment]
@rakyll updated, PTAL

File: stats/view/worker_commands.go [hunk 16]
```diff
@@ -151,7 +155,12 @@ func (cmd *recordReq) handleCommand(w *worker) {
 		}
 		ref := w.getMeasureRef(m.Measure().Name())
 		for v := range ref.views {
-			v.addSample(cmd.tm, m.Value())
+			e := &exemplar.Exemplar{
```
Comment (songy23) [inline stats/view/worker_commands.go]: Creating a new `exemplar` for every `measurement` is very wrong. Going to revert this change.
